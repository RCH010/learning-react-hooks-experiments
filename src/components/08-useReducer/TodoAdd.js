import React from 'react'
import PropTypes from 'prop-types'
import { useForm } from '../../hooks/useForm';

const TodoAdd = ({handleAddTodo}) => {
    const [formValues, handleInputChange, reset] = useForm({
        description: ''
    });

    const getNewID = () => {
        return new Date().getTime();
    }
    
    const handleSubmit = (e) => {
        e.preventDefault();
        if (formValues.description.trim().length <= 1) { return; }

        const newTodo = {
            id: getNewID(),
            desc: formValues.description,
            done: false
        };

        handleAddTodo(newTodo)
        reset();
    }

    return (
        <>
            <h2>Agregar</h2>
            <hr/>

            <form onSubmit={handleSubmit}>
                <input 
                    type='text'
                    name='description'
                    placeholder='Hacer...'
                    autoComplete='off'
                    className='form-control'
                    value={formValues.description}
                    onChange={handleInputChange}
                />
                <button
                    type='submit'
                    className='btn btn-outline-primary btn-block'
                >
                    Agregar
                </button>
            </form>
        </>
    )
}

TodoAdd.propTypes = {
    handleAddTodo: PropTypes.func.isRequired
}

export default TodoAdd
