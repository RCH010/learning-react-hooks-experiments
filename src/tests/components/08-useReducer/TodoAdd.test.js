import { shallow } from 'enzyme';
import React from 'react';
import TodoAdd from '../../../components/08-useReducer/TodoAdd';

describe('Tests in TodoAdd', () => {

    const handleAddTodo = jest.fn();
    const wrapper = shallow(<TodoAdd handleAddTodo={handleAddTodo} />)

    test('should match with snapshot with default props', () => {
        expect(wrapper).toMatchSnapshot();
    })
    
    test('should not call handleAddTodo in form', () => {
        const formsubmit = wrapper.find('form').prop('onSubmit');
        formsubmit({preventDefault(){}});
        expect(handleAddTodo).toHaveBeenCalledTimes(0);
    })
    
    test('should call handleAddTodo and reset form', () => {
        const input = wrapper.find('input');
        const value = {
            target: {
                name: 'description',
                value: 'Hola mundo!'
            }
        }
        input.simulate('change', value);
        const formsubmit = wrapper.find('form').prop('onSubmit');
        formsubmit({preventDefault(){}});
        expect(handleAddTodo).toHaveBeenCalledTimes(1);
        expect(handleAddTodo).toHaveBeenCalledWith(expect.any(Object)); // {}, puede ser un obj vacio y pasaria, hay que ser m√°s precisos
        expect(handleAddTodo).toHaveBeenCalledWith({
            id: expect.any(Number),
            desc: value.target.value,
            done: false
        })
        // expect now that reset has ben called
        expect(wrapper.find('input').prop('value')).toBe('');
    })
    
})
