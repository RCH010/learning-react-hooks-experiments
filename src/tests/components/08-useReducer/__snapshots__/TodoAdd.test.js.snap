// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in TodoAdd should match with snapshot with default props 1`] = `
<Fragment>
  <h2>
    Agregar
  </h2>
  <hr />
  <form
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="form-control"
      name="description"
      onChange={[Function]}
      placeholder="Hacer..."
      type="text"
      value=""
    />
    <button
      className="btn btn-outline-primary btn-block"
      type="submit"
    >
      Agregar
    </button>
  </form>
</Fragment>
`;
